content/courses/advanced-ada/parts/data_types/types.rst

```diff
@@ -231,16 +231,336 @@ In this example, we see instances of the following kinds of names:
+necessarily the case, as we discuss in this section. (In fact, a constant
+object only has a constant view if it doesn't have a part that has a variable
+view.)
+
+A part of an object has a variable view if it is of
+:ref:`immutably limited type <Adv_Ada_Immutably_Limited_Types>`,
+:ref:`controlled type <Adv_Ada_Controlled_Types_Overview_Controlled_Objects>`,
+:ref:`private type <Adv_Ada_Type_View>`, or private extension.
+In that sense, if any of those parts with variable view exist in a constant
+object, then we say that the *whole object* has a variable view.
+Only if a constant object doesn't have *any* parts with variable view, then
+this object has a constant view.
+
+.. todo::
+
+    Add link to section on private extension.
+
+In contrast, variable objects always have a variable view.
+
+Let's see an example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Types.Objects.Object_View
+
+    package Devices is
+
+       type Device_Settings is
+       record
+          Started : Boolean;
+       end record;
+
+       type Device (Id : Positive) is
+         private;
+
+       function Init (Id : Positive)
+                      return Device;
+
+    private
+
+       type Device (Id : Positive) is
+         null record;
+
+       function Init (Id : Positive)
+                      return Device is
+          (Device'(Id => Id));
+
+    end Devices;
+
+
+    with Devices; use Devices;
+
+    procedure Show_Object_View is
+       Dev      : constant Device := Init (5);
+       --  Constant object with
+       --  variable view.
+
+       Default  : constant Device_Settings
+                   := (Started => False);
+       --  Constant object with
+       --  constant view.
+
+       Settings : Device_Settings;
+
+    begin
+       Settings := (Started => True);
+    end Show_Object_View;
+
+In this example, both :ada:`Default_S` and :ada:`Dev` are constant objects.
+However, they have different views: while :ada:`Default_S` has a constant view
+because it doesn't have any parts with variable view, :ada:`Dev` has a variable
+view because it's a private type. Finally, as expected, :ada:`Settings` has a
+variable view because it's a variable object.
+
+
+Named numbers
+~~~~~~~~~~~~~
+
+In addition to objects, we can have named numbers. Those aren't objects, but
+rather :ref:`names <Adv_Ada_Names>` that we assign to numeric values. For
+example:
+
+.. todo::
+
+    Add link to section on names.
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Types.Objects.Named_Number
+
+    procedure Show_Named_Number is
+
+       Pi : constant := 3.1415926535;
+
+    begin
+       null;
+    end Show_Named_Number;
+
+In this example, :ada:`Pi` is a named number.
+
+A named number is always known at compilation time. Also, it doesn't have a
+type associated with it. In fact, its type is called universal real or
+universal integer |mdash| depending on the number being a real or integer
+number.
```

Comment:
I think this needs more info, especially the fact that there is no loss of precision when used with arbitrary types, unlike constants of specific numeric types. I'll send my slides on named numbers to you.